base/timing.jl

```diff
@@ -56,8 +56,17 @@ function gc_alloc_count(diff::GC_Diff)
 end
 
 # cumulative total time spent on compilation
-cumulative_compile_time_ns_before() = ccall(:jl_cumulative_compile_time_ns_before, UInt64, ())
-cumulative_compile_time_ns_after() = ccall(:jl_cumulative_compile_time_ns_after, UInt64, ())
+const comp_timer_lock = ReentrantLock()
+function cumulative_compile_time_ns_before()
+    lock(comp_timer_lock) do
```

Comment:
I don't think this lock makes sense. The right approach is probably to make the flag thread-local.

Reply:
I think you're right. Perhaps the cumulative compile timer counter should be thread local too, to solve the illogical results when `@time` is run on multiple threads (example in the comment above)?

Reply:
Updated to be thread-local. Seems better! Comment below